design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,302 @@
+choose to specify or ignore these fields. To prevent the inconsistencies described above, though,
+we will automatically enforce consistency whenever the key property is set in the `WITH` clause.
+
+* If the user sets a `KEY` property, they do not need to supply `ROWKEY` in their `INSERT INTO`
+statement (e.g. `INSERT INTO foo (id, foo) VALUES ("key", 123);`). In this case, the kafka message
+produced will automatically populate `ROWKEY` to the value of `id`: 
+  ```json
+  {"key": "key", "value": {"id": "key", "foo": 123}}
+  ```
+* If the user sets a `KEY` property _and_ they supply `ROWKEY` in their `INSERT INTO` statement, 
+KSQL will enforce that the value of `ROWKEY` and the value of the key column are identical.
+  ```json
+  {"key": "this_will_fail", "value": {"id": "key", "foo":  123}} 
+  ```
+* If the user does not set a `KEY` property, and they are inserting into a stream, they may choose
+to omit `ROWKEY` altogether. This will result in a `null` key.
+* If the user does not set a `KEY` property, and they are inserting into a table, they _must_
+have a `ROWKEY` value in their `INSERT INTO` statement.
+
+For `ROWTIME`, if it is supplied then that will be the value populated in the kafka record,
+otherwise broker time is used. If a timestamp extractor is also supplied, the `ROWTIME` value will
+still be sent, but the extractor will overwrite the data in `ROWTIME`.
+
+### Complex Datatypes
+
+Although not in scope for V1, complex data types can all be supported following precedents of 
+other SQL-like languages. For example, Google BigQuery supports constructing STRUCT using the
+[syntax](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#declaring-a-struct-type) 
+below (see link for full set of sample syntax):
+
+| Syntax                                    | Output                                              |
+| ----------------------------------------  | --------------------------------------------------- |
+| `STRUCT(1,2,3)`                           | `STRUCT<int64,int64,int64>` *                       |
+| `STRUCT(1 AS a, 'abc' AS b)`              | `STRUCT<a int64, b string>`                         |
+| `STRUCT<x int64>(5)`                      | `STRUCT<x int64>`                                   |
+| `STRUCT(5, STRUCT(5))`                    | `STRUCT<int64, STRUCT<int64>>`                      |
+
+\* we may not want to support "unnamed" structs, thought that decision can be made later
+
+## Test plan
+
+This functionality is expected to work both only in Interactive mode, and these tests must succeed:
+
+* Test that values can be serialized into JSON/AVRO/DELIMITED
+* Test that schemas are properly verified
+* Test `NULL` and table tombstones
+* Test `CREATE STREAM/TABLE` with topics that don't exist
+* Test producing data in topics that already exist
+* Test producing data to topics that are not associated with a KSQL source
```

Comment:
> Test producing data to topics that are not associated with a KSQL source

This isn't possible, right?  You can only INSERT VALUES into a source KSQL knows about....

Reply:
Correct. This should be impossible - I've clarified it in the KLIP.